File: src/test/scala/com/ajjpj/amapper/javabean/propbased/PropertyAccessorTest.scala
package com.ajjpj.amapper.javabean.propbased

import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.matchers.ShouldMatchers
import org.scalatest.FunSuite
import com.ajjpj.amapper.javabean.JavaBeanTypes
import com.ajjpj.amapper.javabean.builder.{NoQualifierExtractor, AMapperExpressionParser}
import com.ajjpj.amapper.core.NoQualifier
import com.ajjpj.amapper.classes.{WithProperties, PartOfPropPath}

/**
 * @author arno
 */
@RunWith(classOf[JUnitRunner])
class PropertyAccessorTest extends FunSuite with ShouldMatchers {
  val cls = classOf[WithProperties]
  val otherCls = classOf[PartOfPropPath]

  def testSimpleAccessor(acc: PropertyAccessor, propName: String = "propName", auxSetter: Option[(WithProperties, String) => Unit] = None) = {
    acc.name should equal (propName)
    acc.isReadable should equal (true)

    auxSetter match {
      case None =>
        acc.isWritable should equal (true)

        val o = new WithProperties
        o.inner.inner = o

        acc.get(o) should equal (null)
        acc.set(o, "xyz")
        acc.get(o) should equal ("xyz")
      case Some(setter) =>
        acc.isWritable should equal (false)

        val o = new WithProperties
        o.inner.inner = o

        acc.get(o) should equal (null)
        setter(o, "xyz")
        acc.get(o) should equal ("xyz")
    }
  }

  def testPathAccessor(acc: PropertyAccessor, propName: String = "propName", firstNullSafe: Boolean, secondNullSafe: Boolean, lastNullSafe: Boolean) = {
    testSimpleAccessor(acc, propName = propName)

    val o = new WithProperties
    if(lastNullSafe)
      acc.get(o) should equal (null)
    else
      evaluating (acc.get(o)) should produce [Exception]

    o.inner = null
    if(secondNullSafe)
      acc.get(o) should equal (null)
    else
      evaluating (acc.get(o)) should produce [Exception]

    if(firstNullSafe)
      acc.get(null) should equal (null)
    else
      evaluating (acc.get(null)) should produce [Exception]
  }

  test ("field") {
    testSimpleAccessor (FieldBasedPropertyAccessor("propName", cls.getDeclaredField("theString"), isDeferred=true, JavaBeanTypes[String], NoQualifier, NoQualifier))
  }

  test ("method") {
    testSimpleAccessor (MethodBasedPropertyAccessor("propName", Some(cls.getDeclaredMethod("getAbc")), Some(cls.getDeclaredMethod("setAbc", classOf[String])), isDeferred=true, JavaBeanTypes[String], NoQualifier, NoQualifier))

    MethodBasedPropertyAccessor("a", Some(cls.getDeclaredMethod("getAbc")), None,                  isDeferred=true, JavaBeanTypes[String], NoQualifier, NoQualifier).isWritable should equal (false)
    MethodBasedPropertyAccessor("a", None, Some(cls.getDeclaredMethod("setAbc", classOf[String])), isDeferred=true, JavaBeanTypes[String], NoQualifier, NoQualifier).isReadable should equal (false)
  }

  test ("ognl") {
    testSimpleAccessor (OgnlPropertyAccessor("propName", "xyz", classOf[WithProperties], isDeferred=true, JavaBeanTypes[String], NoQualifier, NoQualifier))

    OgnlPropertyAccessor("propName", "xyz", classOf[WithProperties], isDeferred=true, JavaBeanTypes[String], NoQualifier, NoQualifier).isWritable should equal (true)
    OgnlPropertyAccessor("readOnly", "readOnly", classOf[WithProperties], isDeferred=true, JavaBeanTypes[String], NoQualifier, NoQualifier).isWritable should equal (false)
  }

  test ("method path") {
    val steps = List(MethodPathStep(cls.getDeclaredMethod("getOther"), nullSafe=false), MethodPathStep(otherCls.getDeclaredMethod("getWithProperties"), nullSafe=false))
    val acc = MethodPathBasedPropertyAccessor("propName", steps, Some(cls.getDeclaredMethod("getAbc")), Some(cls.getDeclaredMethod("setAbc", classOf[String])), finalStepNullSafe = false, isDeferred = false, JavaBeanTypes[String], NoQualifier, NoQualifier)

    testPathAccessor(acc, firstNullSafe = false, secondNullSafe = false, lastNullSafe = false)

    //TODO test readable / writable
  }

  test ("method path first segment null safe") {
    val steps = List(MethodPathStep(cls.getDeclaredMethod("getOther"), nullSafe=true), MethodPathStep(otherCls.getDeclaredMethod("getWithProperties"), nullSafe=false))
    val acc = MethodPathBasedPropertyAccessor("propName", steps, Some(cls.getDeclaredMethod("getAbc")), Some(cls.getDeclaredMethod("setAbc", classOf[String])), finalStepNullSafe = false, isDeferred = false, JavaBeanTypes[String], NoQualifier, NoQualifier)

    testPathAccessor(acc, firstNullSafe = true, secondNullSafe = false, lastNullSafe = false)
  }

  test ("method path middle segment null safe") {
    val steps = List(MethodPathStep(cls.getDeclaredMethod("getOther"), nullSafe=false), MethodPathStep(otherCls.getDeclaredMethod("getWithProperties"), nullSafe=true))
    val acc = MethodPathBasedPropertyAccessor("propName", steps, Some(cls.getDeclaredMethod("getAbc")), Some(cls.getDeclaredMethod("setAbc", classOf[String])), finalStepNullSafe = false, isDeferred = false, JavaBeanTypes[String], NoQualifier, NoQualifier)

    testPathAccessor(acc, firstNullSafe = false, secondNullSafe = true, lastNullSafe = false)
  }

  test ("method path last segment null safe") {
    val steps = List(MethodPathStep(cls.getDeclaredMethod("getOther"), nullSafe=false), MethodPathStep(otherCls.getDeclaredMethod("getWithProperties"), nullSafe=false))
    val acc = MethodPathBasedPropertyAccessor("propName", steps, Some(cls.getDeclaredMethod("getAbc")), Some(cls.getDeclaredMethod("setAbc", classOf[String])), finalStepNullSafe = true, isDeferred = false, JavaBeanTypes[String], NoQualifier, NoQualifier)

    testPathAccessor(acc, firstNullSafe = false, secondNullSafe = false, lastNullSafe = true)
    //TODO test null safety of setter
  }

  test ("parsed method") {
    val acc = new AMapperExpressionParser(NoQualifierExtractor).parse(cls, "theString", JavaBeanTypes[String], isDeferred = false)
    testSimpleAccessor(acc, propName="theString")
  }

  test ("parsed ognl") {
    val acc = new AMapperExpressionParser(NoQualifierExtractor).parse(cls, "1>0?theString:null", JavaBeanTypes[String], isDeferred = false)
    testSimpleAccessor(acc, propName="1>0?theString:null", auxSetter = None)
  }

  test ("parsed path as ognl") {
    val acc = new AMapperExpressionParser(NoQualifierExtractor).parse(cls, "theString.substring(0)", JavaBeanTypes[String], isDeferred = false)
    acc.isInstanceOf[OgnlPropertyAccessor] should equal (true)
  }

  test ("parsed path") {
    val acc = new AMapperExpressionParser(NoQualifierExtractor).parse(cls, "other.withProperties.theString", JavaBeanTypes[String], isDeferred = false)
    testPathAccessor(acc, propName="other.withProperties.theString", firstNullSafe = false, secondNullSafe = false, lastNullSafe = false)
  }

  test ("parsed path null-safe first") {
    val acc = new AMapperExpressionParser(NoQualifierExtractor).parse(cls, "?other.withProperties.theString", JavaBeanTypes[String], isDeferred = false)
    testPathAccessor(acc, propName="?other.withProperties.theString", firstNullSafe = true, secondNullSafe = false, lastNullSafe = false)
  }

  test ("parsed path null-safe second") {
    val acc = new AMapperExpressionParser(NoQualifierExtractor).parse(cls, "other.?withProperties.theString", JavaBeanTypes[String], isDeferred = false)
    testPathAccessor(acc, propName="other.?withProperties.theString", firstNullSafe = false, secondNullSafe = true, lastNullSafe = false)
  }

  test ("parsed path null-safe last") {
    val acc = new AMapperExpressionParser(NoQualifierExtractor).parse(cls, "other.withProperties.?theString", JavaBeanTypes[String], isDeferred = false)
    testPathAccessor(acc, propName="other.withProperties.?theString", firstNullSafe = false, secondNullSafe = false, lastNullSafe = true)
  }
}

File: src/test/java/com/ajjpj/amapper/javabean/PropertyAccessorTest.java
package com.ajjpj.amapper.javabean;

import com.ajjpj.afoundation.collection.immutable.AOption;
import com.ajjpj.amapper.classes.PartOfPropPath;
import com.ajjpj.amapper.classes.WithProperties;
import com.ajjpj.amapper.core.tpe.AQualifier;
import com.ajjpj.amapper.javabean.builder.ABeanExpressionParser;
import com.ajjpj.amapper.javabean.builder.qualifier.NoQualifierExtractor;
import com.ajjpj.amapper.javabean.propbased.accessors.*;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.*;


/**
 * @author arno
 */
public class PropertyAccessorTest {

    public void checkSimpleAccessor(APropertyAccessor acc) throws Exception {
        checkSimpleAccessor(acc, "propName");
    }

    public void checkSimpleAccessor(APropertyAccessor acc, String propName) throws Exception {
        checkSimpleAccessor(acc, propName, AOption.<APropertyAccessor>none());
    }

    public void checkSimpleAccessor(APropertyAccessor acc, String propName, AOption<APropertyAccessor> auxSetter) throws Exception {
        assertEquals(propName, acc.getName());

        final WithProperties o = new WithProperties();
        o.inner.inner = o;

        if(auxSetter.isDefined()) {
            assertEquals(false, acc.isWritable());

            assertEquals(null, acc.get(o));
            auxSetter.get().set(o, "xyz");
            assertEquals("xyz", acc.get(o));
        }
        else {
            assertEquals(true, acc.isWritable());

            assertEquals(null, acc.get(o));
            acc.set(o, "xyz");
            assertEquals("xyz", acc.get(o));
        }
    }

    public void checkPathAccessor(APropertyAccessor acc, boolean firstNullSafe, boolean secondNullSafe, boolean lastNullSafe) throws Exception {
        checkSimpleAccessor(acc, acc.getName());

        final WithProperties o = new WithProperties();
        if(lastNullSafe) {
            assertEquals(null, acc.get(o));
        }
        else {
            try {
                acc.get(o);
                fail("exception expected");
            }
            catch (Exception exc) { //
            }
        }

        o.inner = null;
        if(secondNullSafe) {
            assertEquals(null, acc.get(o));
        }
        else {
            try {
                acc.get(o);
                fail("exception expected");
            }
            catch (Exception exc) { //
            }
        }

        if(firstNullSafe) {
            assertEquals(null, acc.get(null));
        }
        else {
            try {
                acc.get(null);
                fail("exception expected");
            }
            catch (Exception exc) { //
            }
        }
    }

    @Test
    public void testField() throws Exception {
        checkSimpleAccessor(new AFieldBasedPropertyAccessor("propName", WithProperties.class.getDeclaredField("theString"), false, JavaBeanTypes.create(String.class), AQualifier.NO_QUALIFIER, AQualifier.NO_QUALIFIER));
    }

    @Test
    public void testMethod() throws Exception {
        checkSimpleAccessor(new AMethodBasedPropertyAccessor("propName", WithProperties.class.getDeclaredMethod("getAbc"), WithProperties.class.getDeclaredMethod("setAbc", String.class), false, JavaBeanTypes.create(String.class), AQualifier.NO_QUALIFIER, AQualifier.NO_QUALIFIER));
        assertEquals(false, new AMethodBasedPropertyAccessor("a", WithProperties.class.getDeclaredMethod("getAbc"), null, true, JavaBeanTypes.create(String.class), AQualifier.NO_QUALIFIER, AQualifier.NO_QUALIFIER).isWritable());
    }

    @Test
    public void testOgnl() throws Exception {
        checkSimpleAccessor(new AOgnlPropertyAccessor("propName", "xyz", WithProperties.class, true, JavaBeanTypes.create(String.class), AQualifier.NO_QUALIFIER, AQualifier.NO_QUALIFIER));

        assertEquals(true,  new AOgnlPropertyAccessor("propName", "xyz",      WithProperties.class, true, JavaBeanTypes.create(String.class), AQualifier.NO_QUALIFIER, AQualifier.NO_QUALIFIER).isWritable());
        assertEquals(false, new AOgnlPropertyAccessor("readOnly", "readOnly", WithProperties.class, true, JavaBeanTypes.create(String.class), AQualifier.NO_QUALIFIER, AQualifier.NO_QUALIFIER).isWritable());
    }

    private void checkMethodPath(boolean firstNullSafe, boolean secondNullSafe, boolean finalNullSafe) throws Exception {
        final List<AMethodPathBasedPropertyAccessor.Step> steps = new ArrayList<AMethodPathBasedPropertyAccessor.Step>();
        steps.add(new AMethodPathBasedPropertyAccessor.Step(WithProperties.class.getMethod("getOther"), firstNullSafe));
        steps.add(new AMethodPathBasedPropertyAccessor.Step(PartOfPropPath.class.getMethod("getWithProperties"), secondNullSafe));

        final APropertyAccessor acc = new AMethodPathBasedPropertyAccessor("propName", steps, WithProperties.class.getMethod("getAbc"), WithProperties.class.getMethod("setAbc", String.class), finalNullSafe, false, JavaBeanTypes.create(String.class), AQualifier.NO_QUALIFIER, AQualifier.NO_QUALIFIER);

        checkPathAccessor(acc, firstNullSafe, secondNullSafe, finalNullSafe);
    }

    //TODO test readable / writable

    @Test
    public void testMethodPath() throws Exception {
        checkMethodPath(false, false, false);
        checkMethodPath(true,  false, false);
        checkMethodPath(false, true,  false);
        checkMethodPath(false, false, true); //TODO test null safety of setter
    }

    @Test
    public void testParsedMethod() throws Exception {
        final APropertyAccessor acc = new ABeanExpressionParser(NoQualifierExtractor.INSTANCE).parse(WithProperties.class, "theString", JavaBeanTypes.create(String.class), false);
        checkSimpleAccessor(acc, "theString");
    }

    @Test
    public void testParsedOgnl() throws Exception {
        final APropertyAccessor acc = new ABeanExpressionParser(NoQualifierExtractor.INSTANCE).parse(WithProperties.class, "1>0?theString:null", JavaBeanTypes.create(String.class), false);
        checkSimpleAccessor(acc, "1>0?theString:null");
    }

    @Test
    public void testParsedPathAsOgnl() throws Exception {
        final APropertyAccessor acc = new ABeanExpressionParser(NoQualifierExtractor.INSTANCE).parse(WithProperties.class, "theString.substring(0)", JavaBeanTypes.create(String.class), false);
        assertTrue(acc instanceof AOgnlPropertyAccessor);
    }

    @Test
    public void testParsedPath() throws Exception {
        final APropertyAccessor acc = new ABeanExpressionParser(NoQualifierExtractor.INSTANCE).parse(WithProperties.class, "other.withProperties.theString", JavaBeanTypes.create(String.class), false);
        assertTrue(acc instanceof AMethodPathBasedPropertyAccessor);
        checkPathAccessor(acc, false, false, false);
    }

    @Test
    public void testParsedPathFirstNullSafe() throws Exception {
        final APropertyAccessor acc = new ABeanExpressionParser(NoQualifierExtractor.INSTANCE).parse(WithProperties.class, "?other.withProperties.theString", JavaBeanTypes.create(String.class), false);
        assertTrue(acc instanceof AMethodPathBasedPropertyAccessor);
        checkPathAccessor(acc, true, false, false);
    }

    @Test
    public void testParsedPathSecondNullSafe() throws Exception {
        final APropertyAccessor acc = new ABeanExpressionParser(NoQualifierExtractor.INSTANCE).parse(WithProperties.class, "other.?withProperties.theString", JavaBeanTypes.create(String.class), false);
        assertTrue(acc instanceof AMethodPathBasedPropertyAccessor);
        checkPathAccessor(acc, false, true, false);
    }

    @Test
    public void testParsedPathLastNullSafe() throws Exception {
        final APropertyAccessor acc = new ABeanExpressionParser(NoQualifierExtractor.INSTANCE).parse(WithProperties.class, "other.withProperties.?theString", JavaBeanTypes.create(String.class), false);
        assertTrue(acc instanceof AMethodPathBasedPropertyAccessor);
        checkPathAccessor(acc, false, false, true);
    }
}
